[webview_flutter_wkwebview] Extended Web View API on iOS to add flexibility when working with local HTML content

agavrilko opened the pull request:
## Overview

The `readAccessURLProvider` property has been added to the `WebKitWebViewControllerCreationParams` class. This function allows customization of the path to associated resources when loading a local HTML page using `loadFile` on iOS.

### What Is the Issue

Fixes https://github.com/flutter/flutter/issues/136479

The native `WKWebView` takes two arguments when loading local HTML pages:

1. **The file URL** – The local HTML file to be loaded.
2. **The read access URL** – A file or directory that `WKWebView` can access (e.g., .css, .js files).

Currently, the Flutter implementation always sets the parent folder of the specified HTML file as the read access directory. This behavior is not configurable, which limits how local web content can be structured.

For example, the following structure does not work because the .css and .js files are not in the parent directory of the HTML file:
```
/app_resources/
│── styles/
│    ├── main.css
│
│── scripts/
│    ├── app.js
│
│── pages/
     ├── index.html  (Loaded file)
```
With the existing behavior, `WKWebView` cannot `access styles/main.css` and `scripts/app.js` because the parent folder of `index.html` (`/pages/`) does not contain them.

### How This Resolves the Issue

The `readAccessURLProvider` property has been added to the `WebKitWebViewControllerCreationParams` class. This property accepts a function that takes the path of the HTML file being loaded and returns the path to the associated resources.

![Screenshot at Mar 04 14-49-46](https://github.com/user-attachments/assets/be9ce41f-4a8c-4e91-b73c-041e22f59d29)

Each time `loadFile` is called, the controller invokes this function and passes its result as the second argument to the underlying `WKWebView` implementation.

By default, the function returns the parent directory of the specified HTML file, preserving the existing behavior when no custom provider is set.

### Impact on End Users

1. Developers can now explicitly specify the directory that `WKWebView` should allow access to when loading a local HTML page, providing greater flexibility in organizing assets.
2. The existing API signature remains unchanged, ensuring that the update does not require any modifications to existing codebases unless the new functionality is utilized.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

bparrishMines commented:
@agavrilko To follow the pattern of other methods, we should introduce a new platform interface method to `webview_flutter_platform_interface`. Something like `PlatformWebViewController.loadFileWithParams(LoadFileParams)`. `LoadFileParams` would contain a `url` parameter. Then the `webview_flutter_wkwebview` package can create its own implementation of `LoadFileParams` with the `readAccessUrl` parameter because it is unique to iOS/macOS.

bparrishMines reviewed (changes requested):
@agavrilko To follow the pattern of other methods, we should introduce a new platform interface method to `webview_flutter_platform_interface`. Something like `PlatformWebViewController.loadFileWithParams(LoadFileParams)`. `LoadFileParams` would contain a `url` parameter. Then the `webview_flutter_wkwebview` package can create its own implementation of `LoadFileParams` with the `readAccessUrl` parameter because it is unique to iOS/macOS.

agavrilko commented:
Hello @bparrishMines,
Thank you for taking the time to review this MR!
I would like to ensure that I understand your feedback before proceeding.

I initially aimed to introduce this minor update without making any API changes, focusing solely on exposing the functionality that was already present but not accessible to users.
However, based on your comment, it seems you are suggesting a broader update that would introduce new APIs (e.g., `LoadFileParams` and a new/updated platform interface method).

Based on my understanding, your suggestion would require:
1. Introducing `LoadFileParams` at the platform interface level.
2. Extending the `PlatformWebViewController` to accept `LoadFileParams` instead of a plain string.
3. Adapting `webview_flutter_wkwebview` with a `WebKitLoadFileParams` subclass.

![webview_design_update](https://github.com/user-attachments/assets/584156d0-9a4b-41b7-8e5e-229a07cac2bb)

While this design would make the API more consistent and extensible, it would also introduce a breaking change for current users, requiring them to update their code.
That is why I am in doubt.

And if my understanding is correct, should I just update an existing `loadFile` method or introduce a new one?

bparrishMines commented:
@agavrilko I think you should create a new platform interface method named something like: `PlatformWebViewController.loadFileWithParams(LoadFileParams)`. This is consistent with updates to other platform interfaces where we realized in hindsight the method should have taken a class so that we can add parameters later.

Everything else you stated seems correct.

Additionally, after this feature lands, `PlatformWebViewController.loadFile` can be deprecated and it can just call `loadFileWithParams`.

agavrilko commented:
Hello @bparrishMines

I'm seeing that the pipeline fails due to a version resolution issue:
> Because webview_flutter_example depends on webview_flutter_platform_interface ^2.12.0 which doesn't match any versions, version solving failed.

Could you please advise on the proper workflow for updating versions across the packages?
Also, when you have a moment, could you review the changes to ensure everything aligns with your expectations?

Thank you!

bparrishMines commented:
@agavrilko Yes, you can follow https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changing-federated-plugins for the process.

agavrilko commented:
Hello @bparrishMines,

I need further assistance with this PR.
I expected to see `...you can't publish a package that has dependency overrides...`, as stated [here](https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changing-federated-plugins).

But it looks like changing `_platform_interface` and implementations in the same PR is not allowed:
> Changed packages: webview_flutter/webview_flutter,webview_flutter_android,webview_flutter_platform_interface,webview_flutter_wkwebview
> [0:00] Running for packages/webview_flutter/webview_flutter...
> Dart changes are not allowed to other packages in webview_flutter in the same PR as changes to public Dart code in webview_flutter_platform_interface, as this can cause accidental breaking changes to be missed by automated checks. Please split the changes to these two packages into separate PRs.

Should this change be split into 2 different PRs?
If yes, please specify what parts should go to each.

bparrishMines reviewed (commented):
@agavrilko Yes, this PR won't be published as is. The first step is to get this PR approved first, then individual PRs for each package will be created. This allows us to review all the changes together before we start landing seaprate PRs.

agavrilko commented on packages/webview_flutter/webview_flutter/lib/src/webview_controller.dart:
```diff
@@ -131,10 +131,25 @@ class WebViewController {
+  /// such as headers (on Android) or resource paths (on iOS/macOS).
+  ///
+  /// Throws a `PlatformException` if the [absoluteFilePath] does not exist.
+  Future<void> loadFileWithParams(LoadFileParams params) {
```
I am not sure I am property understand this comment.
Do you suggest to remove this method from the `WebViewController` class?
The loading of the file with custom parameters would look like the following:

>  switch (controller.platform) {
>    case final WebKitWebViewController platform:
>      platform.loadFileWithParams(...)
>
>    default:
>      ...
>  }

Just want to ensure I understand it right.

bparrishMines replied:
Correct. `WebViewController.loadFileWithParams` should be removed and users could use the code snippet you provided to access the platform-specific params. The goal of the app-facing package is to provide a simple and clean API while the platform interface provides a more robust and extendable API.

agavrilko replied:
The method was removed.
All changes to the `WebViewController` were reverted.

agavrilko commented on packages/webview_flutter/webview_flutter/lib/src/webview_controller.dart:
```diff
@@ -131,10 +131,25 @@ class WebViewController {
   /// `/Users/username/Documents/www/index.html`.
   ///
   /// Throws a `PlatformException` if the [absoluteFilePath] does not exist.
+  @Deprecated('Use loadFileWithParams(LocalFileParams params) instead. '
```
Updated.

agavrilko commented on packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/load_file_params.dart:
````diff
@@ -0,0 +1,40 @@
+///   final String readAccessPath;
+/// }
+/// ```
+/// {@end-tool}
````
Removed.

agavrilko commented on packages/webview_flutter/webview_flutter_wkwebview/test/webkit_webview_controller_test.dart:
```diff
@@ -330,15 +330,38 @@ void main() {
       });
     });
 
-    test('loadFile', () async {
```
'loadFile' test restored.

agavrilko commented on packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart:
```diff
@@ -38,6 +38,40 @@ enum PlaybackMediaTypes {
+class WebKitLoadFileParams extends LoadFileParams {
+  /// Constructs a [WebKitLoadFileParams], the subclass of a [LoadFileParams].
+  ///
+  /// The optional [readAccessPath] defines the directory that the WebView is
```
Moved.

agavrilko commented on packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:
```diff
@@ -18,6 +18,48 @@ import 'android_webkit_constants.dart';
+class AndroidLoadFileParams extends LoadFileParams {
+  /// Constructs a [AndroidLoadFileParams], the subclass of a [LoadFileParams].
+  ///
+  /// Optionally, [headers] map may be included, which contains key-value pairs
```
Moved.

agavrilko commented on packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:
```diff
@@ -383,15 +425,19 @@ class AndroidWebViewController extends PlatformWebViewController {
       _webView.pigeon_instanceManager.getIdentifier(_webView)!;
 
   @override
-  Future<void> loadFile(
```
Updated.

bparrishMines commented:
@agavrilko Is this waiting for another review? `WebViewController.loadFileWithParams` is still included in `webview_flutter`.

agavrilko commented:
> @agavrilko Is this waiting for another review? `WebViewController.loadFileWithParams` is still included in `webview_flutter`.

Hello @bparrishMines, my apologies, I did not see the notification about your reply.
The method was removed as requested.

I have rebased the branch onto the latest `main` and cleaned the rest of the code.
Let me know what I should do next.
Thank you for your assistance.

agavrilko commented on packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart:
```diff
@@ -61,6 +61,24 @@ abstract class PlatformWebViewController extends PlatformInterface {
+  /// Platform-specific implementations may extend [LocalFileParams] to support
+  /// additional parameters, such as iOS/macOS-specific read access options.
+  ///
+  /// Throws an [ArgumentError] if the file specified in
```
Removed from `loadFileWithParams` only.

agavrilko commented on packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:
```diff
@@ -18,6 +18,40 @@ import 'android_webkit_constants.dart';
+    );
+  }
+
+  /// Additional HTTP headers to be included when loading the local file.
```
Updated.

agavrilko commented on packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:
```diff
@@ -386,12 +420,30 @@ class AndroidWebViewController extends PlatformWebViewController {
+  ) async {
+    switch (params) {
+      case final AndroidLoadFileParams params:
+        await _webView.settings.setAllowFileAccess(true);
```
Updated.
I'm not familiar with the internal implementation, but these operations seem closely related. Could this introduce any potential race conditions? Setting the access flag after the page starts loading doesn't seem reliable to me.

agavrilko commented on packages/webview_flutter/webview_flutter_wkwebview/lib/src/webkit_webview_controller.dart:
```diff
@@ -38,6 +38,38 @@ enum PlaybackMediaTypes {
+    );
+  }
+
+  /// The directory to which the WebView is granted read access.
```
Updated.

agavrilko commented on packages/webview_flutter/webview_flutter_android/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 4.8.0
+
+* Introduces `AndroidLoadFileParams`, a platform-specific extension of `LoadFileParams` for Android that adds support for `headers`.
```
Added.

stuartmorgan-g commented on packages/webview_flutter/webview_flutter/CHANGELOG.md:
Everything in this package can be reverted; there are no changes at this layer.

agavrilko replied:
Changes on `webview_flutter` were reverted.

stuartmorgan-g reviewed (changes requested):
@bparrishMines Longer term we probably want to consider plumbing a platform-subclassable load options class to the main API; having to call the platform interface's load method instead of just passing a platform subclass to an app-facing API method is somewhat awkward, and we generally don't want clients to be in the habit of directly using the platform interface itself.

stuartmorgan-g commented on packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_webview_controller.dart:
```diff
@@ -61,6 +61,21 @@ abstract class PlatformWebViewController extends PlatformInterface {
+  Future<void> loadFileWithParams(
+    LoadFileParams params,
+  ) {
+    throw UnimplementedError(
```
This should call the old method, not throw. See step 1 of https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changing-platform-interface-method-parameters

stuartmorgan-g replied:
And that pass-through should be unit-tested in this package.

agavrilko replied:
`loadFileWithParams` implementation was updated. Unit test added.

stuartmorgan-g reviewed (approved):
Looks good to go forward with splitting out the platform interface PR.

agavrilko commented:
Hello @bparrishMines, @stuartmorgan-g

Here is the PR with only platform interface changes.
[https://github.com/flutter/packages/pull/9697](https://github.com/flutter/packages/pull/9697)

Let me know if I applied the version bump correctly.

agavrilko commented:
Hello @bparrishMines, @stuartmorgan-g 
I updated PR after the platform interface was merged.

All tests and formatting passes just fine locally,
However there is an issue on pipeline, which I am afraid I do not understand:
<img width="617" height="69" alt="image" src="https://github.com/user-attachments/assets/c6281177-6450-4899-b02a-a593cf95f444" />

> error - lib/src/webkit_webview_controller.dart:44:41 - Classes can only extend other classes. Try specifying a different superclass, or removing the extends clause. - extends_non_class
>   error - lib/src/webkit_webview_controller.dart:47:20 - No associated named super constructor parameter. Try changing the name to the name of an existing named super constructor parameter, or creating such named parameter. - super_formal_parameter_without_associated_named
>   error - lib/src/webkit_webview_controller.dart:49:57 - The argument type 'dynamic' can't be assigned to the parameter type 'String'.  - argument_type_not_assignable
>   error - lib/src/webkit_webview_controller.dart:54:5 - Undefined class 'LoadFileParams'. Try changing the name to the name of an existing class, or creating a class with the name 'LoadFileParams'. - undefined_class
>   error - lib/src/webkit_webview_controller.dart:448:5 - Undefined class 'LoadFileParams'. Try changing the name to the name of an existing class, or creating a class with the name 'LoadFileParams'. - undefined_class
>   error - lib/src/webkit_webview_controller.dart:453:18 - The getter 'absoluteFilePath' isn't defined for the type 'WebKitLoadFileParams'. Try importing the library that defines 'absoluteFilePath', correcting the name to the name of an existing getter, or defining a getter or field named 'absoluteFilePath'. - undefined_getter
>   error - test/webkit_webview_controller_test.dart:353:17 - The name 'LoadFileParams' isn't a class. Try correcting the name to match an existing class. - creation_with_non_type
> warning - lib/src/webkit_webview_controller.dart:447:16 - The method doesn't override an inherited method. Try updating this class to match the superclass, or removing the override annotation. - override_on_non_overriding_member
> warning - lib/src/webkit_webview_controller.dart:457:7 - Dead code. Try removing the code, or fixing the code before it so that it can be reached. - dead_code
> warning - lib/src/webkit_webview_controller.dart:458:9 - Dead code. Try removing the code, or fixing the code before it so that it can be reached. - dead_code
>    info - lib/src/webkit_webview_controller.dart:447:16 - Missing documentation for a public member. Try adding documentation for the member. - public_member_api_docs
> 
> 11 issues found.

Looks like the latest version is not being pulled:
> webview_flutter_platform_interface 2.13.0 (2.14.0 available)

But why this happens I can't tell.
Could you guide please?

agavrilko commented:
Thank you @stuartmorgan-g for explaining what `downgrade` tests are.

In this case, since main webview_backage depends on platform interface and depends on wkwebview & andorid, that depend on platform interface as well:

![image](https://github.com/user-attachments/assets/74205ed5-6f44-42f3-9cc7-169b8ba5af48)

I believe it will be right to create another PR to update `webview_flutter_android` and `webview_flutter_wkwebview`.
And then have just `webview_flutter` updated in this one.

Is my understanding correct?

agavrilko commented:
I have update the references in android and wkwebview packages.

agavrilko commented on packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:
```diff
@@ -18,6 +18,41 @@ import 'android_webkit_constants.dart';
+
+  /// Additional HTTP headers to be included when loading the local file.
+  ///
+  /// If not provided at initialization time, doesn't add any addition headers.
```
Updated.
